### PR TITLE
Support /dns protocol in multiaddr

### PR DESCRIPTION
**Pull request by mzabaluev**
The `/dns` protocol has been added to the spec and has had a de-facto meaning for years (citing https://github.com/multiformats/multiaddr/pull/100)

This adds address parsing and encoding support for `/dns` to the multiaddr library.

### CODE REVIEW

**Comment by mzabaluev** [comment]
Address resolution for this is not yet implemented in libp2p, but other users of the `parity-multiaddr` crate could make use of the protocol component in their own ways, e.g. translating a `/dns` multiaddr to an URL to give to a Happy Eyeballs capable HTTP stack.

**Comment by tomaka** [comment]
Support for it is probably very easy to add to `libp2p-dns`.
I don't want to lose track of aa semi-implemented feature, and I don't think it's worth opening an issue just for this, so support should probably be added to this PR, or whoever merges this should quickly submit a PR afterwards.

**Comment by mzabaluev** [comment]
OK, I will try to do here shortly.

**Comment by mzabaluev** [comment]
@tomaka I have added `/dns` support to libp2p-dns and a few other places. I get a unit test failure in libp2p-dns, but the same test is also failing in master.

Commits c02bf8d and possibly f76331f may be too disruptive, as they change the behavior of URL-to-multiaddr translations, albeit improving it with regard to IPv6 support.

File: misc/multiaddr/src/from_url.rs [hunk 1]
```diff
@@ -70,7 +70,7 @@ fn from_url_inner_http_ws(url: url::Url, lossy: bool) -> std::result::Result<Mul
         if let Ok(ip) = hostname.parse::<IpAddr>() {
             Protocol::from(ip)
         } else {
-            Protocol::Dns4(hostname.into())
+            Protocol::Dns(hostname.into())
```
Comment (tomaka) [inline misc/multiaddr/src/from_url.rs:73]: Oh, that's actually a great change :tada:

File: transports/websocket/src/framed.rs [hunk 2]
```diff
@@ -371,7 +373,7 @@ fn location_to_multiaddr<T>(location: &str) -> Result<Multiaddr, Error<T>> {
             let mut a = Multiaddr::empty();
             match url.host() {
                 Some(url::Host::Domain(h)) => {
-                    a.push(Protocol::Dns4(h.into()))
+                    a.push(Protocol::Dns(h.into()))
```
Comment (tomaka) [inline transports/websocket/src/framed.rs:376]: Same

**Pull request merged**